src/azul/service/responseobjects/storage_service.py

```diff
@@ -39,12 +39,137 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
+        self.client.create_bucket(Bucket=(bucket_name or self.bucket_name))
+
+
+class MultipartUploadHandler:
+    """
+    A context manager that facilitates multipart upload to S3. It uploads parts
+    concurrently.
+
+    Sample usage:
+
+    .. code-block:: python
+
+       with MultipartUploadHandler('samples.txt', 'text/plain'):
+           handler.push(b'abc')
+           handler.push(b'defg')
+           # ...
+
+    Upon exit of the body of the with statement, all parts will have been
+    uploaded and the S3 object is guaranteed to exist, or an exception is raised.
+    When an exception is raised within the context, the upload will be aborted
+    automatically.
+    """
+
+    def __init__(self, object_key, content_type: str):
+        self.bucket_name = config.s3_bucket
+        self.object_key = object_key
+        self.upload_id = None
+        self.mp_upload = None
+        self.next_part_number = 1
+        self.content_type = content_type
+        self.parts = []
+        self.futures = []
+        self.thread_pool = None
+
+    def __enter__(self):
+        api_response = boto3.client('s3').create_multipart_upload(Bucket=self.bucket_name,
+                                                                  Key=self.object_key,
+                                                                  ContentType=self.content_type)
+        self.upload_id = api_response['UploadId']
+        self.mp_upload = boto3.resource('s3').MultipartUpload(self.bucket_name, self.object_key, self.upload_id)
+        self.thread_pool = ThreadPoolExecutor(max_workers=MULTIPART_UPLOAD_MAX_WORKERS)
+        return self
+
+    def __exit__(self, etype, value, traceback):
+        if etype:
+            logger.error('Upload %s: Error detected within the MPU context.',
+                         self.upload_id,
+                         exc_info = (etype, value, traceback)
+                         )
+            self.__abort()
+        else:
+            self.__complete()
+
+    def __complete(self):
+        for future in as_completed(self.futures):
+            exception = future.exception()
+            if exception is not None:
+                logger.error('Upload %s: Error detected while uploading a part.',
+                             self.upload_id,
+                             exc_info=(type(exception), exception, None))
+                self.__abort()
+                raise UnexpectedMultipartUploadAbort(f'{self.bucket_name}/{self.object_key}') from exception
+
+        try:
+            self.mp_upload.complete(MultipartUpload={"Parts": [part.to_dict() for part in self.parts]})
+        except self.mp_upload.meta.client.exceptions.ClientError as exception:
+            logger.error('Upload %s: Error detected while completing the upload.',
+                         self.upload_id,
+                         exc_info=(type(exception), exception, None))
+            self.__abort()
+            if 'EntityTooSmall' in exception.args[0]:
```

Comment:
Please eliminate this conditional and the UploadPartSizeOutOfBoundError class. If you use `from …` when raising the exception all the necessary information will be available and there is no need to further differentiate the type of the outer exception.